plugins/InspectionGadgets/InspectionGadgetsAnalysis/src/com/siyeh/ig/bugs/IncorrectDateTimeFormatInspection.java

```diff
@@ -0,0 +1,262 @@
+    }
+
+    @Override
+    public boolean verifier(int count) {
+      return count >= from && count <= to;
+    }
+  }
+
+  private static RangeVerifier rangeOf(int from, int to) {
+    return new RangeVerifier(from, to);
+  }
+
+  private static class SetVerifier implements CountVerifier {
+    private final Set<Integer> set;
+
+    private SetVerifier(int... numbers) {
+      set = Arrays.stream(numbers).boxed()
+        .collect(Collectors.toSet());
+    }
+
+    @Override
+    public boolean verifier(int count) {
+      return set.contains(count);
+    }
+  }
+
+  private static SetVerifier setOf(int... numbers) {
+    return new SetVerifier(numbers);
+  }
+
+  Map<Character, CountVerifier> DATE_TIME_FORMATTER_ALLOWED = Map.ofEntries(
+    Map.entry('G', rangeOf(1, 5)),
+    Map.entry('u', rangeOf(1, 19)),
+    Map.entry('y', rangeOf(1, 19)),
+    Map.entry('Y', rangeOf(1, Integer.MAX_VALUE)), //according to refs must be until 19, but now there is no exception
+    Map.entry('Q', rangeOf(1, 5)),
+    Map.entry('q', rangeOf(1, 5)),
+    Map.entry('M', rangeOf(1, 5)),
+    Map.entry('L', rangeOf(1, 5)),
+    Map.entry('w', rangeOf(1, 2)),
+    Map.entry('W', rangeOf(1, 1)),
+    Map.entry('d', rangeOf(1, 2)),
+    Map.entry('D', rangeOf(1, 3)),
+    Map.entry('F', rangeOf(1, 1)),
+    Map.entry('g', rangeOf(1, 19)),
+    Map.entry('E', rangeOf(1, 5)),
+    Map.entry('e', rangeOf(1, 5)),
+    Map.entry('c', rangeOf(1, 5)),
+
+    Map.entry('a', rangeOf(1, 1)),
+    Map.entry('h', rangeOf(1, 2)),
+    Map.entry('H', rangeOf(1, 2)),
+    Map.entry('k', rangeOf(1, 2)),
+    Map.entry('K', rangeOf(1, 2)),
+    Map.entry('m', rangeOf(1, 2)),
+    Map.entry('s', rangeOf(1, 2)),
+    Map.entry('S', rangeOf(1, 9)), //limit as nanos
+    Map.entry('A', rangeOf(1, 19)),
+    Map.entry('n', rangeOf(1, 19)),
+    Map.entry('N', rangeOf(1, 19)),
+
+    Map.entry('B', setOf(1, 4, 5)),
+
+    Map.entry('V', rangeOf(2, 2)),
+    Map.entry('v', setOf(1, 4)),
+    Map.entry('z', rangeOf(1, 4)),
+    Map.entry('O', setOf(1, 4)),
+    Map.entry('X', rangeOf(1, 5)),
+    Map.entry('x', rangeOf(1, 5)),
+    Map.entry('Z', rangeOf(1, 5))
+  );
+
+  private static final CallMatcher PATTERN_METHODS = anyOf(
+    staticCall("java.time.format.DateTimeFormatter", "ofPattern"),
+    instanceCall("java.time.format.DateTimeFormatterBuilder", "appendPattern").parameterTypes(CommonClassNames.JAVA_LANG_STRING)
+  );
+
+  @NotNull
+  @Override
+  public PsiElementVisitor buildVisitor(@NotNull ProblemsHolder holder, boolean isOnTheFly) {
+    return new JavaElementVisitor() {
+      @Override
+      public void visitMethodCallExpression(@NotNull PsiMethodCallExpression call) {
+        if (PATTERN_METHODS.test(call)) {
+          processExpression(call.getArgumentList().getExpressions()[0]);
+        }
+      }
+
+      private void processExpression(@NotNull PsiExpression expression) {
+        Object patternObject = ConstantExpressionUtil.computeCastTo(expression,
```

Comment:
You may use `com.siyeh.ig.psiutils.ExpressionUtils#computeConstantExpression(com.intellij.psi.PsiExpression)`. You don't need to materialize java.lang.String type there.